File: server/models/booking.js
const mongoose = require("mongoose");
const Schema = mongoose.Schema;
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const config = require("../config");

// For populate
require("./client");
require("./product");
require("./table");

const bookingSchema = new mongoose.Schema(
  {
    price: {
      type: Number,
    },
    products: [
      {
        product: {
          type: Schema.Types.ObjectId,
          ref: "Product",
          required: true,
        },
        quantity: {
          type: Number,
          required: true,
          default: 1,
        },
      },
    ],

    table: {
      type: Schema.Types.ObjectId,
      ref: "Table",
      required: true,
    },

    client: {
      type: Schema.Types.ObjectId,
      ref: "Client",
      required: true,
    },


    profilePhoto: {
      type: String,
    },
  },
  {
    timestamps: true,
  }
);

bookingSchema.methods.toJSON = function () {
  const booking = this;
  const bookingObject = booking.toObject();

  delete bookingObject.createdAt;
  delete bookingObject.updatedAt;

  return bookingObject;
};

bookingSchema.methods.calculatePrice = async function () {
  const booking = this;
  const products = [];
  let total = 0;
  for (let i = 0; i < booking.products.length; i++) {
    const product = await Product.findById(booking.products[i]);    
    total += product.product.price * product.product.price * product.quantity;
  }
  booking.price = total;
  await booking.save();
  return booking.price;
};

const Booking = mongoose.model("Booking", bookingSchema);

module.exports = Booking;

File: lib/models/booking.dart
import 'dart:convert';

import 'package:rinconcito_mexicano/models/client.dart';
import 'package:rinconcito_mexicano/models/product.dart';
import 'package:rinconcito_mexicano/models/table.dart';

class Booking {
    Booking({
        required this.id,
        required this.price,
        required this.products,
        required this.table,
        required this.client,
    });

    final String id;
    final int price;
    final List<ProductElement> products;
    final Table table;
    final Client client;

    factory Booking.fromJson(String str) => Booking.fromMap(json.decode(str));

    String toJson() => json.encode(toMap());

    factory Booking.fromMap(Map<String, dynamic> json) => Booking(
        id: json["_id"],
        price: json["price"],
        products: List<ProductElement>.from(json["products"].map((x) => ProductElement.fromMap(x))),
        table: Table.fromMap(json["table"]),
        client: Client.fromMap(json["client"]),
    );

    Map<String, dynamic> toMap() => {
        "_id": id,
        "price": price,
        "products": List<dynamic>.from(products.map((x) => x.toMap())),
        "table": table.toMap(),
        "client": client.toMap(),
    };
}

class ProductElement {
    ProductElement({
        required this.product,
        required this.quantity,
        required this.id,
    });

    final Product product;
    final int quantity;
    final String id;

    factory ProductElement.fromJson(String str) => ProductElement.fromMap(json.decode(str));

    String toJson() => json.encode(toMap());

    factory ProductElement.fromMap(Map<String, dynamic> json) => ProductElement(
        product: Product.fromMap(json["product"]),
        quantity: json["quantity"],
        id: json["_id"],
    );

    Map<String, dynamic> toMap() => {
        "product": product.toMap(),
        "quantity": quantity,
        "_id": id,
    };
}